cou/steps/plan.py

```diff
@@ -52,24 +53,163 @@
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot determine the current OS release
+        or Ubuntu series.
+    """
+    current_os_release = analysis_result.current_cloud_os_release
+    current_series = analysis_result.current_cloud_series
+    if not current_os_release:
+        raise NoTargetError(
+            "Cannot determine the current OS release in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+    if not current_series:
+        raise NoTargetError(
+            "Cannot determine the current Ubuntu series in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+
+def is_supported_series(analysis_result: Analysis) -> None:
+    """Check the Ubuntu series of the cloud to see if it is supported.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises OutOfSupportRange: When series is not supported.
+    """
+    current_series = analysis_result.current_cloud_series
+    supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
+    # series already checked at is_valid_openstack_cloud
+    if current_series not in supporting_lts_series:  # type: ignore
+        raise OutOfSupportRange(
+            f"Cloud series '{current_series}' is not a Ubuntu LTS series supported by COU. "
+            f"The supporting series are: {supporting_lts_series}"
+        )
+
+
+def is_highest_release_achieved(analysis_result: Analysis) -> None:
+    """Check if the highest OpenStack release is reached by the ubuntu series.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises HighestReleaseAchieved: When the OpenStack release is the last supported by the series.
+    """
+    # series and current OS release already checked at is_valid_openstack_cloud
+    current_os_release = analysis_result.current_cloud_os_release
+    current_series = analysis_result.current_cloud_series
+    if str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]:  # type: ignore
+        raise HighestReleaseAchieved(
+            f"No upgrades available for OpenStack {str(current_os_release).capitalize()} on "
+            f"Ubuntu {current_series.capitalize()}.\nNewer OpenStack releases "  # type: ignore
+            "may be available after upgrading to a later Ubuntu series."
+        )
+
+
+def is_target_supported(analysis_result: Analysis) -> None:
+    """Check if the target to upgrade is supported.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot find target to upgrade
+    :raises OutOfSupportRange: When the upgrade scope is not supported
+        by the current series.
+    """
+    current_os_release = analysis_result.current_cloud_os_release
+    current_series = analysis_result.current_cloud_series
+    if current_os_release and current_series:
```

Comment:
We do not need to check this or do we? It was checked in `is_valid_openstack_cloud`.